src/modules/dashboard/components/Whitelist/Whitelist.tsx

```diff
@@ -1,22 +1,82 @@
 import isEmpty from 'lodash/isEmpty';
-import React from 'react';
+import React, { useCallback } from 'react';
 
+import { defineMessages } from 'react-intl';
+
+import Button from '~core/Button';
+import { useDialog } from '~core/Dialog';
+import { MiniSpinnerLoader } from '~core/Preloaders';
+
+import {
+  useWhitelistAgreementQuery,
+  useLoggedInUser,
+  Colony,
+} from '~data/index';
+import { useTransformer } from '~utils/hooks';
+
+import { getAllUserRoles } from '../../../transformers';
+import { hasRoot } from '../../../users/checks';
+
+import AgreementDialog from './AgreementDialog';
+import styles from './Whitelist.css';
 import UploadAddressesWidget from './UploadAddressesWidget';
 import WhitelistAddresses from './WhitelistAddresses';
 
-export interface Props {
-  colonyAddress: string;
+const MSG = defineMessages({
+  agreement: {
+    id: 'dashboard.Extensions.WhitelisExtension.agreement',
+    defaultMessage: 'Agreement',
+  },
+});
+
+interface Props {
+  colony: Colony;
 }
 
-const Whitelist = ({ colonyAddress }: Props) => {
+const Whitelist = ({ colony: { colonyAddress }, colony }: Props) => {
+  const { walletAddress, username, ethereal } = useLoggedInUser();
+  const hasRegisteredProfile = !!username && !ethereal;
+  const allUserRoles = useTransformer(getAllUserRoles, [colony, walletAddress]);
+  const userHasPermission = hasRegisteredProfile && hasRoot(allUserRoles);
+
   const users = []; // @TODO: Connect with real added users
 
+  const openAgreementDialog = useDialog(AgreementDialog);
+
+  const { data, loading } = useWhitelistAgreementQuery({
+    variables: { colonyAddress },
+  });
```

Comment:
On a second note, I think we'll be better served by moving the agreement loading **inside** the modal.

So something like this:
- always show the agreement link
- when clicked open the agreement modal with a loading spinner
- attempt to load the agreement data
- if successful remove the spinner and show the data
- if failed, show an error message

I think this approach is better because:
- it fixes the above issue I encountered
- it doesn't front-load the agreement data, and makes the page load faster _(ie: it only loads it when it needs it)_

Reply:
All done. It is a better approach indeed. However I didn't use the `getUseApproval` function as Alex explained that it's if the colony has KYC policy included. Alex said that `getAgreementHash` is the way to identify if there is an agreement included.